iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+    RetryLater(anyhow::Error),
+}
+
+/// Future of a get request.
+type GetFut = LocalBoxFuture<'static, Result<(), FailureAction>>;
+
+/// Trait modelling performing a single request over a connection. This allows for IO-less testing.
+pub trait Getter {
+    /// Type of connections the Getter requires to perform a download.
+    type Connection;
+    /// Return a future that performs the download using the given connection.
+    fn get(&mut self, kind: DownloadKind, conn: Self::Connection) -> GetFut;
+}
+
+/// Concurrency limits for the [`Downloader`].
+#[derive(Debug)]
+pub struct ConcurrencyLimits {
+    /// Maximum number of requests the service performs concurrently.
+    pub max_concurrent_requests: usize,
+    /// Maximum number of requests performed by a single peer concurrently.
+    pub max_concurrent_requests_per_peer: usize,
+    /// Maximum number of open connections the service maintains.
+    pub max_open_connections: usize,
+}
+
+impl Default for ConcurrencyLimits {
+    fn default() -> Self {
+        // these numbers should be checked against a running node and might depend on platform
+        ConcurrencyLimits {
+            max_concurrent_requests: 50,
+            max_concurrent_requests_per_peer: 4,
+            max_open_connections: 100,
+        }
+    }
+}
+
+impl ConcurrencyLimits {
+    /// Checks if the maximum number of concurrent requests has been reached.
+    fn at_requests_capacity(&self, active_requests: usize) -> bool {
+        active_requests >= self.max_concurrent_requests
+    }
+
+    /// Checks if the maximum number of concurrent requests per peer has been reached.
+    fn peer_at_request_capacity(&self, active_peer_requests: usize) -> bool {
+        active_peer_requests >= self.max_concurrent_requests_per_peer
+    }
+
+    /// Checks if the maximum number of connections has been reached.
+    fn at_connections_capacity(&self, active_connections: usize) -> bool {
+        active_connections >= self.max_open_connections
+    }
+}
+
+/// Download requests the [`Downloader`] handles.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum DownloadKind {
+    /// Download a single blob entirely.
+    Blob {
+        /// Blob to be downloaded.
+        hash: Hash,
+    },
+    /// Download a collection entirely.
+    Collection {
+        /// Blob to be downloaded.
+        hash: Hash,
+    },
+}
+
+impl DownloadKind {
+    /// Get the requested hash.
+    const fn hash(&self) -> &Hash {
+        match self {
+            DownloadKind::Blob { hash } | DownloadKind::Collection { hash } => hash,
+        }
+    }
+
+    /// Get the ranges this download is requesting.
+    // NOTE: necessary to extend downloads to support ranges of blobs ranges of collections.
+    #[allow(dead_code)]
+    fn ranges(&self) -> RangeSpecSeq {
+        match self {
+            DownloadKind::Blob { .. } => RangeSpecSeq::from_ranges([RangeSet2::all()]),
+            DownloadKind::Collection { .. } => RangeSpecSeq::all(),
```

Comment:
what's the difference between these two `RangeSpecSeq`, the read is if they are the same for the uninitiated 😅

Reply:
the first one means the range `all` just once (so a blob), whereas the second one means the `all` range in the whole span

Reply:
fixed in 12c63dd5

Reply:
I see, thanks